test/e2e/helpers.go

```diff
@@ -464,26 +464,35 @@ func isAzureMachinePoolWindows(amp *infrav1exp.AzureMachinePool) bool {
 
 // getProxiedSSHClient creates a SSH client object that connects to a target node
 // proxied through a control plane node.
-func getProxiedSSHClient(controlPlaneEndpoint, hostname, port string) (*ssh.Client, error) {
+func getProxiedSSHClient(controlPlaneEndpoint, hostname, port string, ioTimeout time.Duration) (*ssh.Client, error) {
 	config, err := newSSHConfig()
 	if err != nil {
 		return nil, err
 	}
 
 	// Init a client connection to a control plane node via the public load balancer
-	lbClient, err := ssh.Dial("tcp", fmt.Sprintf("%s:%s", controlPlaneEndpoint, port), config)
+	c, err := net.DialTimeout("tcp", fmt.Sprintf("%s:%s", controlPlaneEndpoint, port), config.Timeout)
 	if err != nil {
 		return nil, errors.Wrapf(err, "dialing public load balancer at %s", controlPlaneEndpoint)
 	}
+	err = c.SetDeadline(time.Now().Add(ioTimeout))
+	if err != nil {
+		return nil, errors.Wrapf(err, "setting timeout for connection to public load balancer at %s", controlPlaneEndpoint)
+	}
+	conn, chans, reqs, err := ssh.NewClientConn(c, fmt.Sprintf("%s:%s", controlPlaneEndpoint, port), config)
+	if err != nil {
+		return nil, errors.Wrapf(err, "connecting to public load balancer at %s", controlPlaneEndpoint)
+	}
+	lbClient := ssh.NewClient(conn, chans, reqs)
 
 	// Init a connection from the control plane to the target node
-	c, err := lbClient.Dial("tcp", fmt.Sprintf("%s:%s", hostname, port))
+	c, err = lbClient.Dial("tcp", fmt.Sprintf("%s:%s", hostname, port))
```

Comment:
Why don't we want to `DialTimeout` here as well? Not a pain point?

Reply:
This is `(*ssh.Client).Dial` whereas the other one is akin to `net.Dial`. `*ssh.Client` doesn't have a `DialTimeout` method. And that timeout is only enforced when establishing the TCP connection, but we were seeing communication along that connection stall out afterwards, so the real fix here is the addition of the `SetDeadline` call.